library/gdalframework/src/main/java/it/geosolutions/imageio/gdalframework/GDALUtilities.java

```diff
@@ -764,6 +767,81 @@ public static ColorModel buildColorModel(final SampleModel sampleModel) {
         return colorModel;
     }
 
+    /**
+     * Extracts the bands from the provided {@link ImageReadParam},
+     * with EnhancedImageReadParam.getBands() taking precedence
+     * over ImageReadParam.getSourceBands() and overriding it.
+     *
+     * @param imageReadParam
+     * @return
+     */
+    public static int[] extractBands(ImageReadParam imageReadParam) {
+        int[] srcBands = imageReadParam.getSourceBands();
+        if (imageReadParam instanceof EnhancedImageReadParam) {
+            // Let's check if the EnhancedImageReadParam has been set and it contains
+            // a bands array. In that case, this will take precedence over the srcBands
+            final EnhancedImageReadParam eparam = (EnhancedImageReadParam) imageReadParam;
+            if (eparam.getBands() != null) {
+                srcBands = eparam.getBands();
+                imageReadParam.setSourceBands(srcBands);
+            }
+        }
+        return srcBands;
+    }
+
+    /**
+     * Extracts a {@link ColorModel} given a source ColorModel, a
+     * provided {@link SampleModel} and a number of destination bands.
+     * If the number of bands in the provided {@link SampleModel} matches
+     * the destination number of bands, the source color model is returned.
+     *
+     * @param cm
+     *                the original ColorModel.
+     * @param sm
+     *                the SampleModel to be used as reference.
+     * @param destNumBands
+     *                the number of bands to be used for extracting the ColorModel.
+     * @return a new ColorModel based on the provided parameters.
+     */
+    public static ColorModel extractColorModel(ColorModel cm, SampleModel sm, int destNumBands) {
+        int dataType = sm.getDataType();
+        if (sm.getNumBands() != destNumBands) {
+            if (destNumBands == 1) {

```

Comment:
destNumBads == 2 with gray alpha too maybe?

Wondering, the various color models are going to work fine even with float/double types?